### PR TITLE
Add tcmalloc support to the tikv_alloc crate

### PR DESCRIPTION
**Pull request by ZhangHanDong**
Add tcmalloc support to the tikv_alloc crate for issues [#4191](https://github.com/tikv/tikv/issues/4191)

- remove the `no-jemalloc` feature from tikv's `Cargo.toml`
- change the tikv_alloc crate to `default-features = false`
- add `tikv_alloc/jemalloc` to tikv's default features. 
- enabling tcmalloc for tikv means passing `cargo build --no-default-features --features="tikv_alloc/tcmalloc".

### CODE REVIEW

**Comment by hicqu** [comment]
Thanks for your contribution. To resolve the DCO failure, please use `git commit -s` to commit changes. And please run `make format` to format your code.
BTW your work station is based on mac? Could you show `ulimit -n` result with us?

**Comment by ZhangHanDong** [comment]
@hicqu ok

**Comment by ZhangHanDong** [comment]
@hicqu

- `cargo test -- test integrations` result: 

`test result: ok. 273 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out`

- `make format` have been runned.

**Comment by ZhangHanDong** [comment]
@hicqu 

`cargo test --test integrations --no-default-features --features="tikv_alloc/tcmalloc`   result:

```
test result: ok. 273 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

**Comment by brson** [comment]
Thank you very much for the PR @ZhangHanDong! This is exactly what we're looking for. I see a few things I have comments about in the patch, so I'll leave additional feedback against the code itself.

**Comment by brson** [comment]
OK, I left a few comments, but there's more to say about conditional compilation and the goals of this crate's interface.

This crate presents a consistent view to its users of the memory allocator. That is, no matter which allocator is selected, the user gets the same API. This particularly applies to all the extra profiling functions that are implemented for jemalloc, but not other allocators (at least not yet). So when jemalloc is turned off, and if tcmalloc is turned on all those profiling functions are there, but they are no-ops.

It _looks_ to me that in the current patch those functions are only exported when jemalloc is on: the `pub use self::imp::*;` line is deleted, replaced with `pub use self::jemalloc::*;`. The non-jemalloc `imp` module is still _compiled_ but not used.

I'm a bit surprised that didn't show up as a build error, but I suspect it will if tikv is built with the `mem-profiling` feature, which turns on at least some of the code that accesses these extra profiling functions. So as it stands I'd expect `cargo check --all --no-default-features --features="tikv_alloc/tcmalloc, mem-profiling" to not compile.

Now, let's think about in which scenarios we want which allocators:

- As it stands, we want to only turn on jemalloc or tcmalloc on unix, so that's why so many cfgs mention "unix".
- When neither jemalloc or tcmalloc is enabled we want to explicitly turn on the system allocator
- When we're not on unix we want to explicitly turn on the system allocator
- When we are fuzzing we want to use the system allocator

From these lines:

```
#[cfg(all(
    unix,
    not(fuzzing),
    not(feature = "jemalloc"),
    not(feature = "tcmallocator")
))]
```

it looks to me that on window we are not explicitly enabling the system allocator. (In practice the allocator will _probably_ be the system allocator because that's the default, but I'd feel more comfortable if tikv_alloc always set the allocator to something).

So I'd like you to try a few different things here:

First, as you've extracted the `imp` module for jemalloc to it's own file, please extract the default `imp` module to `default.rs`. With both those extracted you can then eliminate the `imp` modules, so the implementation-specific code resides directly in the implementation-specific modules.

Then, instead of exporting the jemalloc module and its implementation, modify the code to always compile _some_ implementation and reexport its contents. This is a pretty common pattern for compile-time extraction: here's a trimmed down example:

```
#[cfg(jemalloc)]
#[path = "jemalloc.rs"]
mod imp;

#[cfg(not(jemalloc))]
#[path = "default.rs"]
mod imp;;

pub use imp::*:
```

Obviously those cfg's aren't quite right, but that's the pattern. That way there is always an internal mod `imp` (which are all assumed to conform to the same interface), and the contents are always exported.

Onto the second thing.

With respect to the cfg's here I only see 1 case where I think the cfgs do something different than I want, and that's with enabling the System allocator - this patch disables the explicit system allocator on Windows.

One nice property of the existing code is that the conditionals, while there are quite a lot of them, are either the same conditional, or the inverse of that same conditional. That's easy to reason about and relatively easy to maintain. I wonder if we can maintain the same property.

Before this patch, here's what the cfg looks like:

```
#[cfg(all(unix, not(fuzzing), not(feature = "no-jemalloc")))]
```

Which says "do this if we're using jemalloc, with all the corresponding conditions that are allowed when using jemalloc (unix, not(fuzzing))". And it can be inverted with

```
#[cfg(not(all(unix, not(fuzzing), not(feature = "no-jemalloc"))))]
```

So with those two we're saying, at compile time, "if jemalloc" or "if not jemalloc". Very elegant. Now we need to be able to say "if jemalloc", "if tcmalloc", and "if not jemalloc and not tcmalloc". There's also the case to think about if the user enables both jemalloc and tcmalloc - I think though that the error resulting from the duplicate global allocator declarations is sufficient that we don't need to worry about it though.

Unfortunately I can't think of another arrangement that is particularly cleaner than what you've already written, with the exception that I think the cfg around the system allocator is incorrect, as is the one around the "default" `imp` module

If we change that case, then in summary we have the following cfg patterns used in this crate:

- `#[cfg(all(unix, not(fuzzing), feature = "jemalloc"))]` - "using jemalloc"
- `#[cfg(all(unix, not(fuzzing), feature = "tcmalloc"))]` - "using tcmalloc"
- `#[cfg(not(all(unix, not(fuzzing), feature = "jemalloc")))]` "not using jemalloc" - this is for default imp module that should be shared between tcmalloc and the system allocator. Note this again puts the `tot` around the entire "using jemalloc" cfg. This captures the windows case and the fuzzing case.
- `#[cfg(not(all(unix, not(fuzzing), any(feature = "jemalloc", feature = "tcmalloc"))))]` - "not using jemalloc or tcmalloc"

Give this a read and see if these make sense to you. I think that one of these should be on every conditionally-compiled statement in lib.rs

---

OK, so now I know this is a massive amount of feedback. Let me try to summarize what I think should be changed:

- Use the crate names in cfgs: "jemallocator", "tcmalloc".
- Don't bother defining the "tcmallocator" feature - users will either use "jemalloc" or "tcmalloc".
- Move the default imp module to a `default.rs` file, as you've done with `jemalloc.rs`.
- Remove the extra 'imp' modules that are now in their own files.
- Always export `pub use imp::*`, so that the API surface is the same no matter which allocator.

That's the big stuff. Also note whatever comments I left individually.

So @ZhangHanDong take this in and see what you think of it. If you agree then please make the requested changes, otherwise let me know what you think.

After you've got this patch in shape, and before landing it, we'll also want to run some benchmarks to see how tikv w/ jemalloc compares to tikv w/ tcmalloc. But don't worry about that yet; let's focus on the patch first.

Thanks! And sorry for the huge review.

**Comment by brson** [comment]
Thanks for the reviews @hicqu.

**Comment by ZhangHanDong** [comment]
@brson thanks for your review. I will try to modify it according to your opinion.

File: components/tikv_alloc/Cargo.toml [hunk 1]
```diff
@@ -27,3 +30,8 @@ optional = true
 features = ["unprefixed_malloc_on_supported_platforms"]
 version = "0.1.9"
 
+[dependencies.tcmalloc]
+optional = true
+# version = "0.1.3"
+git="https://github.com/jmcomets/tcmalloc-rs.git"
```
[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply components/tikv_alloc/Cargo.toml]: 0.1.3  isn't latest , it used overdue api,  look  [https://github.com/jmcomets/tcmalloc-rs/pull/2](https://github.com/jmcomets/tcmalloc-rs/pull/2)

File: Cargo.toml [hunk 2]
```diff
@@ -6,12 +6,12 @@ keywords = ["KV", "distributed-systems", "raft"]
...
 
 [features]
-default = []
+default = ["tikv_alloc/jemalloc"]
 portable = ["rocksdb/portable"]
 sse = ["rocksdb/sse"]
 mem-profiling = ["tikv_alloc/mem-profiling"]
 no-fail = ["fail/no_fail"]
-no-jemalloc = ["tikv_alloc/no-jemalloc"]
+# no-jemalloc = ["tikv_alloc/no-jemalloc"]
```
[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply Cargo.toml]: ok

File: components/tikv_alloc/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply components/tikv_alloc/Cargo.toml]: maybe request the author to post a new version

File: components/tikv_alloc/Cargo.toml [hunk 3]
```diff
@@ -6,14 +6,17 @@ authors = ["Brian Anderson <andersrb@gmail.com>"]
...
+tcmallocator=["tcmalloc"]
 
 # Build jemalloc's profiling features. Without this
 # certain profile functions will return nothing.
 mem-profiling = ["jemallocator/profiling"]
 
-# Don't build jemalloc; use the system allocator.
-no-jemalloc = []
+# # Don't build jemalloc; use the system allocator.
+# no-jemalloc = []
```
[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply components/tikv_alloc/Cargo.toml]: ok

File: components/tikv_alloc/src/jemalloc.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
+pub use self::imp::*;
+
+// The implementation of this crate when jemalloc is turned on
+#[cfg(all(unix, not(fuzzing), feature = "jemalloc"))]
```
[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply components/tikv_alloc/src/jemalloc.rs]: ok

File: components/tikv_alloc/src/lib.rs [hunk 5]
```diff
@@ -40,8 +40,9 @@
 //! - mem-profiling - compiles jemalloc and this crate with profiling
 //!   capability
 //!
-//! - no-jemalloc - compiles without jemalloc, such that TiKV
-//!   will use the system allocator
+//! - jemalloc - compiles default jemalloc
```
[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply components/tikv_alloc/src/lib.rs]: ok

File: components/tikv_alloc/src/lib.rs [hunk 6]
```diff
@@ -40,8 +40,9 @@
...
+//!
+//! - tcmalloc - compiles with `cargo build --no-default-features --features="tikv_alloc/tcmalloc"`
```
[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply components/tikv_alloc/src/lib.rs]: i see.

File: components/tikv_alloc/src/lib.rs [hunk 7]
```diff
@@ -88,200 +89,45 @@
...
+use libc;
+
 #[cfg(feature = "mem-profiling")]
 #[macro_use]
 extern crate log;
 
-// The global allocator, usually.
-#[cfg(all(unix, not(fuzzing), not(feature = "no-jemalloc")))]
+// TCMalloc
+#[cfg(all(unix, not(fuzzing), feature = "tcmallocator"))]
```
[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply components/tikv_alloc/src/lib.rs]: oh, there is not use crate's name for conditional features name .

[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply components/tikv_alloc/src/lib.rs]: can i use "je" and "tc" for feature name ?

[the remark this answers is not in the document]
Comment (ZhangHanDong) [reply components/tikv_alloc/src/lib.rs]: Features and dependencies cannot have the same name: `tcmalloc`

**Comment by ZhangHanDong** [comment]
@brson

I have been refacted the code according by your opinion.  Test is ok， please review again.

but i have little confused, I have to set `not(feature="jemalloc")` like this:

```rust
#[cfg(all(unix, not(fuzzing), not(feature="jemalloc"), feature = "tcmalloc"))]
```

Otherwise, an error will be reported when build

```
$  cargo test --no-default-features --features="tikv_alloc/tcmalloc"

   Compiling tikv_alloc v0.1.0 (/HOME/work/projects/rust/tikv/components/tikv_alloc)
error[E0428]: the name `ALLOC` is defined multiple times
   --> components/tikv_alloc/src/lib.rs:120:1
    |
112 | static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
    | -------------------------------------------------------------- previous definition of the value `ALLOC` here
...
120 | static ALLOC: TCMalloc = tcmalloc::TCMalloc;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `ALLOC` redefined here
    |
    = note: `ALLOC` must be defined only once in the value namespace of this module

error: cannot define more than one #[global_allocator]
   --> components/tikv_alloc/src/lib.rs:120:1
    |
120 | static ALLOC: TCMalloc = tcmalloc::TCMalloc;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0428`.
error: Could not compile `tikv_alloc`.
```

**Comment by ZhangHanDong** [comment]
@brson

`tcmalloc` author published the new 0.2.0 version for Rust Stable 1.28+ .   I have used 0.2.0 in tikv_alloc crate.

**Comment by brson** [comment]
@ZhangHanDong thank you for the quick turnaround.

I've spent some time looking into the problem you found where you needed to insert `not(jemalloc)`. It's a tough one.

The reason for this is that the defaults are not being turned off. So the code is seeing both `jemalloc` and `tcmalloc` features.

Unfortunately I can't even figure out how to make the tikv crate turn off its default features with `--no-default-features`. It seems to just not do anything

Here's a tree with some experimients: https://github.com/brson/tikv/tree/tcmalloc

I also tried to reproduce the behavior in a standalone test case: https://github.com/brson/featuretest2. It doesn't exhibit the same behavior, `--no-default-features` works.

Maybe a fresh set of eyes can figure something out. Otherwise, I'm a bit stumped. Might need to spend some time stripping down tikv to a minimum-reproducible test case.

**Comment by ZhangHanDong** [comment]
@brson I test your experimental tree,  `--no-default-features` doesn't work.

image

I am in the macOS 10.13.6.

**Comment by brson** [comment]
@ZhangHanDong ok, after thinking about this issue some more, the reason that `--no-default-features` isn't turning off tikv's default features is because there are other projects in the workspace that depend on tikv, and _they_ don't turn off the default features.

So the way to make `--no-default-features` work is to add `default-features = false` to each crate that links to tikv's Cargo.toml. Those look to be fuzz/targets, components/test_coprocessor, components/test_raftstore, components/test_storage, and components/test_util.

If those are all modified then it should be possible to turn off tikv's default features and make this feature scheme work. I think.

**Comment by ZhangHanDong** [comment]
@brson   It's ok now,  I also modify each crate links to tikv_alloc.

but I want to close this pr, re-pull request, to solve the problem of DOC

**Comment by ZhangHanDong** [comment]
@brson Re-PR :  [https://github.com/tikv/tikv/pull/4370](https://github.com/tikv/tikv/pull/4370)

**Pull request closed without merging**